File: lesson-3_dynamic-programming/0-1-knapsack-problem/recurrence/main.mjs
import { stdin as input, stdout as output } from 'node:process'
import * as readline from 'node:readline/promises'

const rl = readline.createInterface({ input, output })

let tmp = (await rl.question('')).split(' ')
const n = tmp[0]
const balance = tmp[1]
const values = []
const like = []
const fav = Array.from({ length: balance + 1 }, () => 0)
for (let i = 0; i < n; i++) {
    tmp = (await rl.question('')).split(' ')
    values[i] = Number(tmp[0])
    like[i] = Number(tmp[1])
}
for (let i = 1; i <= balance; i++) {
    if (values[0] <= i) {
        fav[i] = like[0];
    }
}
for (let i = 1; i < n; i++) {
    for (let j = balance; j >= 1; j--) {
        if (values[i] <= j) {
            fav[j] = Math.max(fav[j], fav[j - values[i]] + like[i]);
        }
    }
}
console.log(fav[balance])
rl.close()

File: lesson-3_dynamic-programming/0-1-knapsack-problem/recurrence/main.kt
import kotlin.math.max

fun main() {
    var tmp = readLine()!!.split(' ')
    val n = tmp[0].toInt()
    val balance = tmp[1].toInt()
    val values = Array(n) { 0 }
    val like = Array(n) { 0 }
    val fav = Array(balance + 1) { 0 }
    for (i in 0 until n) {
        tmp = readLine()!!.split(' ')
        values[i] = tmp[0].toInt()
        like[i] = tmp[1].toInt()
    }
    for (i in 1..balance) {
        if (values[0] <= i) {
            fav[i] = like[0]
        }
    }
    for (i in 1 until n) {
        for (j in balance downTo 1) {
            if (values[i] <= j) {
                fav[j] = max(fav[j], fav[j - values[i]] + like[i])
            }
        }
    }
    println(fav[balance])
}
